homeassistant/components/weather/__init__.py

```diff
@@ -207,93 +531,200 @@ def precision(self) -> float:
+                pressure_f = float(pressure)
+                value_pressure = UNIT_CONVERSIONS[ATTR_PRESSURE_UOM](
+                    pressure_f, from_unit, to_unit
                 )
-            data[ATTR_WEATHER_PRESSURE] = pressure
+                data[ATTR_WEATHER_PRESSURE] = round(value_pressure, ROUNDING_PRECISION)
+                data[ATTR_WEATHER_PRESSURE_UNIT] = to_unit
 
         if (wind_bearing := self.wind_bearing) is not None:
             data[ATTR_WEATHER_WIND_BEARING] = wind_bearing
 
-        if (wind_speed := self.wind_speed) is not None:
-            if (unit := self.wind_speed_unit) is not None:
-                wind_speed = round(
-                    self.hass.config.units.wind_speed(wind_speed, unit),
-                    ROUNDING_PRECISION,
+        if (wind_speed := (self.native_wind_speed or self.wind_speed)) is not None:
+            if (native_wind_speed_unit := self.native_wind_speed_unit) is not None:
+                from_unit = native_wind_speed_unit
+                to_unit = self.wind_speed_unit
+            else:
+                from_unit = self.wind_speed_unit
+                to_unit = self.hass.config.units.wind_speed_unit
+            with suppress(ValueError):
+                wind_speed_f = float(wind_speed)
+                value_wind_speed = UNIT_CONVERSIONS[ATTR_WIND_SPEED_UOM](
+                    wind_speed_f, from_unit, to_unit
                 )
-            data[ATTR_WEATHER_WIND_SPEED] = wind_speed
-
-        if (visibility := self.visibility) is not None:
-            if (unit := self.visibility_unit) is not None:
-                visibility = round(
-                    self.hass.config.units.length(visibility, unit), ROUNDING_PRECISION
+                data[ATTR_WEATHER_WIND_SPEED] = round(
+                    value_wind_speed, ROUNDING_PRECISION
+                )
+                data[ATTR_WEATHER_WIND_SPEED_UNIT] = to_unit
+
+        if (visibility := (self.native_visibility or self.visibility)) is not None:
+            if (native_visibility_unit := self.native_visibility_unit) is not None:
+                from_unit = native_visibility_unit
+                to_unit = self.visibility_unit
+            else:
+                from_unit = self.visibility_unit
+                to_unit = self.hass.config.units.length_unit
+            with suppress(ValueError):
+                visibility_f = float(visibility)
+                value_visibility = UNIT_CONVERSIONS[ATTR_VISIBILITY_UOM](
+                    visibility_f, from_unit, to_unit
                 )
-            data[ATTR_WEATHER_VISIBILITY] = visibility
+                data[ATTR_WEATHER_VISIBILITY] = round(
+                    value_visibility, ROUNDING_PRECISION
+                )
+                data[ATTR_WEATHER_VISIBILITY_UNIT] = to_unit
 
         if self.forecast is not None:
             forecast = []
             for forecast_entry in self.forecast:
                 forecast_entry = dict(forecast_entry)
```

Comment:
We don't want `native_*` ending up in the state attribute, maybe do like this + adjust below
```suggestion
            for native_forecast_entry in self.forecast:
                forecast_entry = {}
```